lib/SILGen/SILGenDestructor.cpp

```diff
@@ -213,6 +213,99 @@ void SILGenFunction::emitDeallocatingDestructor(DestructorDecl *dd) {
   B.createReturn(loc, emitEmptyTuple(loc));
 }
 
+bool Lowering::needsIsolatingDestructor(DestructorDecl *dd) {
+  auto ai = swift::getActorIsolation(dd);
+  if (!ai.isActorIsolated()) {
+    return false;
+  }
+  DestructorDecl *firstIsolated = dd;
+  while (true) {
+    DestructorDecl *next = firstIsolated->getSuperDeinit();
+    if (!next)
+      break;
+    auto ai = swift::getActorIsolation(next);
+    if (!ai.isActorIsolated())
+      break;
+    firstIsolated = next;
+  }
+
+  // If isolation was intoduced in ObjC code, then we assume that ObjC code also
+  // overrides retain/release to make sure that dealloc is called on the correct
+  // executor in the first place.
+  return firstIsolated->getClangNode().isNull();
```

Comment:
I think there's an additional optimization we should consider here, although I'm not sure how tricky it is. When the deinitializer is for a "default actor", i.e., one that has not overridden `unownedExecutor`, we shouldn't need to hop to the executor because we know it's available. I wonder if we could contrive to immediately run the deallocating initializer rather than going through "perform on executor." The only thing I'm not sure of is how to ensure that `swift_task_isCurrentExecutor` will return true; perhaps we can temporarily set the thread-specific storage to the actor's executor, run the deallocating initializer code, then set it back.

Reply:
Implemented, please take a look